proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+`prioritization fees` +
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+Where PRAW fees is :
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+If the payer does not have enough balance, the transaction is not scheduled and
+fails with an error `InsufficientFunds`. Consider this case the same as if the
+payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+PRAW fees. For all accounts with PRAW fees, the fees paid should be greater than
+the required fees. In case of overpayment, the difference is stored in a
+variable and will be paid back to the payer in any case. In case the PRAW fees
+are insufficiently paid or not paid then the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The PRAW fees minus rebates and minus overpayment are transferred to the
+respective accounts lamport balance, from where the owner can collect them and
+then the rest would be transferred back to the payer.
+
+### Syscall to change PRAW fees
+
+PRAW fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.The runtime will automatically set the PRAW fee to
+`0` on account creation. PRAW fee will not be changed if the ownership of the
+account changes.
+
+A new syscall will be added to update the PRAW fees:
+
+```
+fn set_program_rebatable_account_write_fee(
+  address: Pubkey, 
+  microlamports_per_requested_cu: u64
+) -> Result<()>;
+```
+
+The syscall will take as arguments the account on which fee should be updated
+and the microlamports per cu that should be charged and return a `Ok` if the
+changes were successful else will return `Err`.
+
+The accounts database will need to store the value of microlamports per
+requested CU for the account. Changes to update the PRAW fees could be done only
+by the account owner. All programs will need to implement special instructions
+so that the program admin can use this syscall to update PRAW fees on the
+required accounts.
+
+Setting the microlamports per CU to `0` disables PRAW fees. The maximum amount
+of microlamports per CU that could be set will be 2 ^ 42. Lets assume that the
+maximum CU required to update these fees on an account is 1,000 CU. So the
+maximum PRAW fees required to update an account (`M`) is
+
+```
+M = 2^42 * 1000 CU * 10^-6 lamports
+M = 4.398046511104 * 10^12 * 1000 * 10^-6
+M = 4.398046511104 SOLs
+```
+
+So this makes sure that admin does not lock itself out from changing the PRAW
+fees.
+
+### Syscall to read PRAW fees
+
+A new syscall will be added to read PRAW fees on an account.
+
+```
+fn get_program_rebatable_account_write_fee(
```

Comment:
maybe this getter syscall can be called by other programs? or forbid for now? (i don't see any practice use case by other programs...)

Reply:
No use for other programs actually.

Reply:
also, for this condition (https://github.com/solana-foundation/solana-improvement-documents/pull/16/files#r1243174480) not honored, how about returning `-> Result<u64>`?

Reply:
I mean `get_program_rebatable_account_write_fee` will always return the PRAW fee associated with the account, even if it is not owned by the calling program.  No use for other programs actually, but it would not hurt to let other programs get this data if they want to do some additional checks etc..